contracts/oracles/CurveOracle.sol

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.11;
+pragma abicoder v1;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "../interfaces/IOracle.sol";
+import "../libraries/Sqrt.sol";
+
+import "../interfaces/ICurveRegistry.sol";
+import "../interfaces/ICurveProvider.sol";
+import "../interfaces/ICurveSwap.sol";
+
+contract CurveOracle is IOracle {
+    using Sqrt for uint256;
+
+    ICurveProvider public immutable addressProvider;
+    IERC20 private constant _NONE = IERC20(0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF);
+
+    constructor(ICurveProvider _addressProvider) {
+        addressProvider = _addressProvider;
+    }
+
+    function getRate(IERC20 _srcToken, IERC20 _dstToken, IERC20 connector) external view override returns (uint256 rate, uint256 weight) {
+
+        require(connector == _NONE, "connector should be NONE");
+
+        ICurveRegistry registry;
+        address pool;
+
+        // first check crypto registry (curve v2) for pools and select first one
+        registry = ICurveRegistry(addressProvider.get_address(5));
+        pool = registry.find_pool_for_coins(address(_srcToken), address(_dstToken));
+
+        // if not found, check main registry and select first one
+        if (pool == address(0)){
+            registry = ICurveRegistry(addressProvider.get_address(0));
+            pool = registry.find_pool_for_coins(address(_srcToken), address(_dstToken));
+        }
+
+        // if not found, check factory registry (permissionless) and select first one
+        if (pool == address(0)){
+            registry = ICurveRegistry(addressProvider.get_address(3));
+            pool = registry.find_pool_for_coins(address(_srcToken), address(_dstToken));
+        }
+
+        require(pool != address(0), "Pool does not exist");
+
+        int128 a;
+        int128 b;
+        bool underlying;
+        uint256[8] memory balances;
+        uint256[8] memory decimals;
+        (a, b, underlying) = registry.get_coin_indices(pool, address(_srcToken), address(_dstToken));
+        decimals = registry.get_decimals(pool);
+
+        // this is for converting int128 -> uint
+        int128 j = 0;
+        uint256 srcBalance;
+        uint256 dstBalance;
+        uint256 srcDecimal;
+        for (uint256 i = 0; i < 8; i++) {
+            if (j == a) {
+                srcBalance = balances[i];
+                srcDecimal = decimals[i];
+            } else if (j == b) {
+                dstBalance = balances[i];
+            }
+            j++;
+        }
+        weight = (srcBalance * dstBalance).sqrt();
+
+        if (!underlying) {
+            rate = ICurveSwap(pool).get_dy(a, b, 10 ** srcDecimal) * 1e12;
```

Comment:
The `get_dy(i, j, dx)` and `get_dy_underlying(i, j, dx)` returns how many `j`-tokens you will get if you will give `dx` amount of `i`-token. So rate is not eqauls to result of these methods. 
I think you have not to use result for 1 token, we have to use result for all balance to calculate the rate.